module/mempool/stdmap/backdata/herocache/cache.go

```diff
@@ -0,0 +1,403 @@
+}
+
+func NewCache(limit uint32, oversizeFactor uint32, ejectionMode heropool.EjectionMode, logger zerolog.Logger) *Cache {
+	// total buckets.
+	capacity := uint64(limit * oversizeFactor)
+	bucketNum := capacity / slotsPerBucket
+	if capacity%slotsPerBucket != 0 {
+		// accounting for remainder.
+		bucketNum++
+	}
+
+	bd := &Cache{
+		logger:                 logger,
+		bucketNum:              bucketNum,
+		limit:                  limit,
+		buckets:                make([]slotBucket, bucketNum),
+		ejectionMode:           ejectionMode,
+		entities:               heropool.NewHeroPool(limit, ejectionMode),
+		availableSlotHistogram: make([]uint64, slotsPerBucket+1), // +1 is to account for empty buckets as well.
+	}
+
+	return bd
+}
+
+// Has checks if we already contain the entity with the given identifier.
+func (c *Cache) Has(entityID flow.Identifier) bool {
+	defer c.logTelemetry()
+
+	_, _, _, ok := c.get(entityID)
+	return ok
+}
+
+// Add adds the given item to the BackData.
+func (c *Cache) Add(entityID flow.Identifier, entity flow.Entity) bool {
+	defer c.logTelemetry()
+
+	return c.put(entityID, entity)
+}
+
+// Rem will remove the item with the given identity.
+func (c *Cache) Rem(entityID flow.Identifier) (flow.Entity, bool) {
+	defer c.logTelemetry()
+
+	entity, bucketIndex, sliceIndex, exists := c.get(entityID)
+	if !exists {
+		return nil, false
+	}
+	// removes value from underlying entities list.
+	c.invalidateEntity(bucketIndex, sliceIndex)
+
+	// frees up slot
+	c.deallocateSlot(bucketIndex, sliceIndex)
+
+	return entity, true
+}
+
+// Adjust will adjust the value item using the given function if the given identifier can be found.
+// Returns a bool which indicates whether the value was updated as well as the updated value.
+func (c *Cache) Adjust(entityID flow.Identifier, f func(flow.Entity) flow.Entity) (flow.Entity, bool) {
+	defer c.logTelemetry()
+
+	entity, removed := c.Rem(entityID)
+	if !removed {
+		return nil, false
+	}
+
+	newEntity := f(entity)
+	newEntityID := newEntity.ID()
+
+	c.put(newEntityID, newEntity)
+
+	return newEntity, true
+}
+
+// ByID returns the given entity from the BackData.
+func (c *Cache) ByID(entityID flow.Identifier) (flow.Entity, bool) {
+	defer c.logTelemetry()
+
+	entity, _, _, ok := c.get(entityID)
+	return entity, ok
+}
+
+// Size will return the total size of BackData, i.e., total number of valid (entityId, entity) pairs.
+func (c Cache) Size() uint {
+	defer c.logTelemetry()
+
+	return uint(c.entities.Size())
+}
+
+// All returns all entities from the BackData.
+func (c Cache) All() map[flow.Identifier]flow.Entity {
+	defer c.logTelemetry()
+
+	all := make(map[flow.Identifier]flow.Entity)
+	for bucketIndex, bucket := range c.buckets {
+		for slotIndex := range bucket {
+			id, entity, linked := c.linkedEntityOf(bIndex(bucketIndex), sIndex(slotIndex))
+			if !linked {
+				// slot may never be used, or recently invalidated
+				continue
+			}
+
+			all[id] = entity
+		}
+	}
+
+	return all
```

Comment:
> This seems very heavy weight to convert all cached entities into a Golang map. I'm guessing this is for backwards compatibility with the old cache implementation?

Yes

> I'm curious about how the callers use the map once it's returned to them. Is there a way to optimize this mass instantiation and copying of memory away somehow? Should there be an issue to investigate this further, later? Or now, if we fear it might cause performance issues?

Returned map used solely to iterate on, hence issued up here to fix https://github.com/dapperlabs/flow-go/issues/6102